src/python/pants/engine/rules.py

```diff
@@ -21,6 +21,14 @@
 from pants.util.ordered_set import FrozenOrderedSet, OrderedSet
 
 
+class CanModifyWorkunit:
```

Comment:
For improved typechecking, you can make this an abstract class with an abstract method to return the level.

Naming wise: how about `EngineAwareResult`, `WorkunitResult`, `DynamicResult`...?